### PR TITLE
Fix SSM gradient transforms across chunks

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Falcon-H1 fine-tuning crashed because `ssmAttn` forced evaluation while MLX was tracing a gradient transformation.

This removes the internal evaluation barriers so gradients remain connected across SSM chunks and adds a regression test covering chunked gradient computation.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx-swift-lm/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by davidkoski** [comment]
I am curious what the underlying issue is here - could it be related to https://github.com/ml-explore/mlx-swift/pull/461?

Do we have a vulnerability in general when doing fine tuning + eval with the async evals?  I wonder if we need a _synchronous only_ bit on `GenerateParameters`?

Or perhaps this is just eval-ish code inside something that is going to run through compile (this seems more likely looking at where this code lives)?

**Comment by davidkoski** [comment]
Reading the description more closely it isn't compile (but the same issue) it is `grad()` or `valueAndGrad()`.

This change looks good.

I wonder if we should consider something in mlx-swift with a task local that would either allow `eval` / `asyncEval` to NOP in these cases?  `item()` can't be called, but the eval variants should be.  They could be treated as advisory under grad/compile.

**Review by davidkoski — APPROVED** [review]
Looks good to me.

**Comment by aleroot** [comment]
> I am curious what the underlying issue is here - could it be related to [ml-explore/mlx-swift#461](https://github.com/ml-explore/mlx-swift/pull/461)?
> 
> Do we have a vulnerability in general when doing fine tuning + eval with the async evals? I wonder if we need a _synchronous only_ bit on `GenerateParameters`?
> 
> Or perhaps this is just eval-ish code inside something that is going to run through compile (this seems more likely looking at where this code lives)?

#461 and this share the tracing/evaluation surface, but they fix different failures. #461 fixes a cross-thread lock inversion while here `ssmAttn` was directly materializing tracer arrays inside grad/valueAndGrad. A synchronous generation flag would not fix this because synchronous eval is also invalid on tracers, and training does not use GenerateParameters. I agree there may be room in mlx-swift for an explicitly advisory evaluation hint that no-ops during transforms, but I would keep ordinary eval strict.

**Comment by aleroot** [comment]
> Reading the description more closely it isn't compile (but the same issue) it is `grad()` or `valueAndGrad()`.
> 
> This change looks good.
> 
> I wonder if we should consider something in mlx-swift with a task local that would either allow `eval` / `asyncEval` to NOP in these cases? `item()` can't be called, but the eval variants should be. They could be treated as advisory under grad/compile.

See [this](https://github.com/ml-explore/mlx-swift-lm/pull/574)
